File: crates/kglite-mcp-server/src/tools.rs
//! KGLite-specific MCP tools: `cypher_query`, `graph_overview`, `save_graph`.
//!
//! All three close over a [`GraphState`] holding the active
//! [`kglite::api::KnowledgeGraph`] behind an `Arc<RwLock<…>>`. Wired
//! into the framework's tool router via `register_typed_tool` so they
//! sit alongside the built-in source / GitHub tools.
//!
//! 0.9.18: rewritten against the pure-Rust `kglite::api` surface.
//! There is no `Python::attach` anywhere in this module — the binary
//! has no libpython link at all.

use std::path::Path;
use std::sync::{Arc, RwLock};

use anyhow::Result;
use kglite::api::cypher;
use kglite::api::{
    compute_description, compute_schema, load_file, ConnectionDetail, CypherDetail, Embedder,
    FluentDetail, KnowledgeGraph, Value,
};
use mcp_methods::server::McpServer;
use serde::{Deserialize, Serialize};

const NO_GRAPH: &str =
    "No active graph. Pass --graph X.kgl, or activate one via repo_management('org/repo').";

/// Shared active-graph state. Cloning is cheap (Arc).
#[derive(Clone, Default)]
pub struct GraphState {
    inner: Arc<RwLock<Option<ActiveGraph>>>,
}

struct ActiveGraph {
    kg: KnowledgeGraph,
    source_path: Option<std::path::PathBuf>,
}

impl GraphState {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn load_kgl(&self, path: &Path) -> Result<()> {
        let kg = load_file(&path.to_string_lossy())
            .map_err(|e| anyhow::anyhow!("kglite::load_file failed: {}", e))?;
        *self.inner.write().unwrap() = Some(ActiveGraph {
            kg,
            source_path: Some(path.to_path_buf()),
        });
        Ok(())
    }

    pub fn build_code_tree(&self, dir: &Path) -> Result<()> {
        let kg = kglite::api::build_code_tree(dir, false, true, None, None)
            .map_err(|e| anyhow::anyhow!("kglite::build_code_tree failed: {}", e))?;
        *self.inner.write().unwrap() = Some(ActiveGraph {
            kg,
            source_path: None,
        });
        Ok(())
    }

    pub fn bind_embedder(&self, embedder: Arc<dyn Embedder>) -> Result<()> {
        let mut guard = self.inner.write().unwrap();
        let Some(active) = guard.as_mut() else {
            tracing::warn!("embedder loaded before any graph is active; binding deferred");
            return Ok(());
        };
        active.kg.set_embedder_native(embedder);
        Ok(())
    }

    pub fn schema(&self) -> Option<(u64, u64)> {
        let guard = self.inner.read().unwrap();
        let active = guard.as_ref()?;
        let overview = compute_schema(active.kg.dir());
        Some((overview.node_count as u64, overview.edge_count as u64))
    }

    fn with_active<F>(&self, f: F) -> String
    where
        F: FnOnce(&ActiveGraph) -> String,
    {
        let guard = self.inner.read().unwrap();
        match guard.as_ref() {
            Some(active) => f(active),
            None => NO_GRAPH.to_string(),
        }
    }

    /// Resolve a code-entity qualified name to its source location via
    /// `KnowledgeGraph::source_location`. Used by the `read_code_source`
    /// tool to bridge the qualified-name → file path lookup.
    pub fn source_lookup(
        &self,
        qualified_name: &str,
        node_type: Option<&str>,
    ) -> Result<crate::code_source::SourceLookup, String> {
        let guard = self.inner.read().unwrap();
        let Some(active) = guard.as_ref() else {
            return Err(NO_GRAPH.to_string());
        };
        match active.kg.source_location(qualified_name, node_type) {
            kglite::api::SourceLookup::Found(loc) => {
                let file_path = loc.file_path.ok_or_else(|| {
                    format!("graph.source({qualified_name:?}) returned no file_path")
                })?;
                let line_number = loc.line_number.unwrap_or(1).max(1) as usize;
                let end_line = loc.end_line.unwrap_or(loc.line_number.unwrap_or(1)).max(1) as usize;
                Ok(crate::code_source::SourceLookup {
                    file_path,
                    line_number,
                    end_line,
                })
            }
            kglite::api::SourceLookup::Ambiguous(matches) => Err(format!(
                "ambiguous qualified_name {qualified_name:?}; matches: {matches:?}. \
                 Pass `node_type` to narrow."
            )),
            kglite::api::SourceLookup::NotFound => Err(format!(
                "graph.source({qualified_name:?}) returned no match. \
                 Try passing `node_type` or using a different qualified name."
            )),
        }
    }

    /// Run a parameterised Cypher template against the active graph.
    /// Used by the YAML-declared `tools[].cypher` registration path
    /// (see [`crate::cypher_tools::register_cypher_tools`]).
    pub fn run_cypher_template(
        &self,
        template: &str,
        args: &serde_json::Map<String, serde_json::Value>,
    ) -> String {
        let guard = self.inner.read().unwrap();
        let Some(active) = guard.as_ref() else {
            return NO_GRAPH.to_string();
        };
        let mut params = std::collections::HashMap::new();
        for (k, v) in args {
            params.insert(k.clone(), json_to_value(v));
        }
        match run_cypher_inner(&active.kg, template, params) {
            Ok(body) => body,
            Err(e) => format!("Cypher error: {e}"),
        }
    }
}

/// Convert a `serde_json::Value` into a Cypher param `Value`. Mirrors
/// the Python boundary's `py_value_to_value` for the JSON subset.
fn json_to_value(v: &serde_json::Value) -> Value {
    match v {
        serde_json::Value::Null => Value::Null,
        serde_json::Value::Bool(b) => Value::Boolean(*b),
        serde_json::Value::Number(n) => {
            if let Some(i) = n.as_i64() {
                Value::Int64(i)
            } else if let Some(f) = n.as_f64() {
                Value::Float64(f)
            } else {
                Value::Null
            }
        }
        serde_json::Value::String(s) => Value::String(s.clone()),
        // Arrays and objects flow through as JSON-serialised strings; the
        // Cypher engine doesn't have a first-class list/map Value variant
        // at the param boundary, so this matches the existing behaviour
        // for non-scalar JSON inputs from MCP tool calls.
        other => Value::String(other.to_string()),
    }
}

/// Run a Cypher query against the given KnowledgeGraph snapshot. Picks
/// between read and write paths based on `is_mutation_query`; on success
/// returns the rendered tool body (CSV when `FORMAT CSV` is in the
/// query, inline 15-row preview otherwise).
fn run_cypher_inner(
    kg: &KnowledgeGraph,
    query: &str,
    params: std::collections::HashMap<String, Value>,
) -> Result<String, String> {
    let mut parsed =
        cypher::parse_cypher(query).map_err(|e| format!("Cypher syntax error: {e}"))?;
    let mut params = params;

    let rewrite = cypher::rewrite_text_score(&mut parsed, &params)?;
    if !rewrite.texts_to_embed.is_empty() && !parsed.explain {
        let model = kg
            .embedder()
            .ok_or_else(|| {
                "text_score() requires a registered embedding model. \
                 Configure `extensions.embedder:` in the manifest."
                    .to_string()
            })?
            .clone();
        model.load()?;
        let texts: Vec<String> = rewrite
            .texts_to_embed
            .iter()
            .map(|(_, t)| t.clone())
            .collect();
        let embeddings = model.embed(&texts);
        model.unload();
        let embeddings = embeddings?;
        if embeddings.len() != texts.len() {
            return Err(format!(
                "text_score: model.embed() returned {} vectors for {} texts",
                embeddings.len(),
                texts.len()
            ));
        }
        for (i, (param_name, _)) in rewrite.texts_to_embed.iter().enumerate() {
            let json = format!(
                "[{}]",
                embeddings[i]
                    .iter()
                    .map(|f| f.to_string())
                    .collect::<Vec<_>>()
                    .join(", ")
            );
            params.insert(param_name.clone(), Value::String(json));
        }
    }

    cypher::planner::optimize_with_disabled(
        &mut parsed,
        kg.dir(),
        &params,
        cypher::planner::empty_disabled_set(),
    );
    cypher::mark_lazy_eligibility(&mut parsed);

    let output_csv = parsed.output_format == cypher::OutputFormat::Csv;

    let result = if cypher::is_mutation_query(&parsed) {
        // Mutation requires &mut DirGraph; the shim holds an
        // Arc<DirGraph> snapshot via the KnowledgeGraph, so we cannot
        // mutate it here without a write lock. The shim today only
        // exposes mutating queries when explicitly invoked through
        // `cypher_query`, and the surrounding RwLock<Option<ActiveGraph>>
        // is held read-only across the call. For 0.9.18 we forbid
        // mutations through the MCP tool surface — agents that need to
        // edit the graph use a CLI shell instead.
        return Err(
            "mutation Cypher (CREATE/SET/DELETE/REMOVE/MERGE) is not allowed through \
             the MCP cypher_query tool. Use the kglite CLI for graph edits."
                .to_string(),
        );
    } else {
        let executor = cypher::CypherExecutor::with_params(kg.dir(), &params, None);
        executor
            .execute(&parsed)
            .map_err(|e| format!("Cypher execution error: {e}"))?
    };

    if output_csv {
        Ok(result.to_csv())
    } else {
        Ok(format_cypher_inline(&result))
    }
}

/// Render a CypherResult as an inline 15-row preview (header + repr per
/// row). Matches the format the pre-0.9.18 Python shim produced via
/// `format_cypher_result`.
fn format_cypher_inline(result: &cypher::CypherResult) -> String {
    let len = result.rows.len();
    if len == 0 {
        return "No results.".to_string();
    }
    let header = if len > 15 {
        format!("{len} row(s) (showing first 15):\n")
    } else {
        format!("{len} row(s):\n")
    };
    let mut out = header;
    out.push_str(&result.columns.join("\t"));
    out.push('\n');
    for row in result.rows.iter().take(15) {
        for (i, val) in row.iter().enumerate() {
            if i > 0 {
                out.push('\t');
            }
            push_value_repr(&mut out, val);
        }
        out.push('\n');
    }
    out
}

fn push_value_repr(out: &mut String, val: &Value) {
    use std::fmt::Write;
    match val {
        Value::Null => out.push_str("null"),
        Value::String(s) => {
            let _ = write!(out, "{s:?}");
        }
        Value::Int64(n) => {
            let _ = write!(out, "{n}");
        }
        Value::Float64(f) => {
            let _ = write!(out, "{f}");
        }
        Value::Boolean(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::UniqueId(u) => {
            let _ = write!(out, "{u}");
        }
        Value::DateTime(d) => out.push_str(&d.format("%Y-%m-%d").to_string()),
        Value::Point { lat, lon } => {
            let _ = write!(out, "POINT({lon} {lat})");
        }
        Value::Duration {
            months,
            days,
            seconds,
        } => {
            let _ = write!(out, "duration(M={months}, D={days}, S={seconds})");
        }
        Value::NodeRef(idx) => {
            let _ = write!(out, "node[{idx}]");
        }
    }
}

#[derive(Debug, Default, Deserialize, Serialize, schemars::JsonSchema)]
struct CypherArgs {
    /// Cypher query string. Append `FORMAT CSV` for CSV-encoded output.
    pub query: String,
}

#[derive(Debug, Default, Deserialize, Serialize, schemars::JsonSchema)]
struct OverviewArgs {
    /// Drill into specific node types (e.g. `["Person", "Document"]`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub types: Option<Vec<String>>,
    /// `true` for all connection types; or `["CALLS"]` for a deep-dive.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub connections: Option<serde_json::Value>,
    /// `true` for the Cypher language reference; or `["MATCH","WHERE"]`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cypher: Option<serde_json::Value>,
}

#[derive(Debug, Default, Deserialize, Serialize, schemars::JsonSchema)]
struct SaveGraphArgs {}

/// Builtins toggled by the manifest's `builtins:` block.
#[derive(Clone, Copy, Debug, Default)]
pub struct Builtins {
    pub save_graph: bool,
    pub temp_cleanup_on_overview: bool,
}

pub fn register(server: &mut McpServer, state: GraphState, builtins: Builtins) {
    let s = state.clone();
    server.register_typed_tool::<CypherArgs, _>(
        "cypher_query",
        "Run a Cypher query against the active knowledge graph. Returns up to 15 rows \
         inline; append FORMAT CSV to export full results to a CSV string.",
        move |args| s.with_active(|g| run_cypher_tool(g, &args.query)),
    );
    let s = state.clone();
    let cleanup_temp = builtins.temp_cleanup_on_overview;
    server.register_typed_tool::<OverviewArgs, _>(
        "graph_overview",
        "Inspect the active graph's schema. With no args returns the inventory; pass \
         types=[...] / connections=true|[...] / cypher=true|[...] for drill-down.",
        move |args| {
            if cleanup_temp
                && args.types.is_none()
                && args.connections.is_none()
                && args.cypher.is_none()
            {
                wipe_temp_dir();
            }
            s.with_active(|g| run_overview(g, &args))
        },
    );
    if builtins.save_graph {
        let s = state;
        server.register_typed_tool::<SaveGraphArgs, _>(
            "save_graph",
            "Persist the active graph to its source .kgl file (single-graph mode only).",
            move |_| s.with_active(run_save),
        );
    }
}

fn wipe_temp_dir() {
    let dir = std::path::Path::new("temp");
    if !dir.is_dir() {
        return;
    }
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(e) => {
            tracing::debug!(error = %e, "temp_cleanup: read_dir failed");
            return;
        }
    };
    for entry in entries.flatten() {
        let path = entry.path();
        let res = if path.is_dir() {
            std::fs::remove_dir_all(&path)
        } else {
            std::fs::remove_file(&path)
        };
        if let Err(e) = res {
            tracing::debug!(path = %path.display(), error = %e, "temp_cleanup: remove failed");
        }
    }
}

fn run_cypher_tool(graph: &ActiveGraph, query: &str) -> String {
    match run_cypher_inner(&graph.kg, query, std::collections::HashMap::new()) {
        Ok(s) => s,
        Err(e) => format!("Cypher error: {e}"),
    }
}

fn run_overview(graph: &ActiveGraph, args: &OverviewArgs) -> String {
    let conn = parse_connection_detail(args.connections.as_ref());
    let cy = parse_cypher_detail(args.cypher.as_ref());
    let fluent = FluentDetail::Off;
    match compute_description(
        graph.kg.dir(),
        args.types.as_deref(),
        &conn,
        &cy,
        &fluent,
        None,
        None,
        None,
    ) {
        Ok(s) => s,
        Err(e) => format!("graph_overview error: {e}"),
    }
}

fn parse_connection_detail(v: Option<&serde_json::Value>) -> ConnectionDetail {
    use serde_json::Value;
    match v {
        None | Some(Value::Null) => ConnectionDetail::Off,
        Some(Value::Bool(false)) => ConnectionDetail::Off,
        Some(Value::Bool(true)) => ConnectionDetail::Overview,
        Some(Value::Array(items)) => {
            let names: Vec<String> = items
                .iter()
                .filter_map(|i| i.as_str().map(String::from))
                .collect();
            if names.is_empty() {
                ConnectionDetail::Overview
            } else {
                ConnectionDetail::Topics(names)
            }
        }
        Some(_) => ConnectionDetail::Overview,
    }
}

fn parse_cypher_detail(v: Option<&serde_json::Value>) -> CypherDetail {
    use serde_json::Value;
    match v {
        None | Some(Value::Null) => CypherDetail::Off,
        Some(Value::Bool(false)) => CypherDetail::Off,
        Some(Value::Bool(true)) => CypherDetail::Overview,
        Some(Value::Array(items)) => {
            let names: Vec<String> = items
                .iter()
                .filter_map(|i| i.as_str().map(String::from))
                .collect();
            if names.is_empty() {
                CypherDetail::Overview
            } else {
                CypherDetail::Topics(names)
            }
        }
        Some(_) => CypherDetail::Overview,
    }
}

fn run_save(graph: &ActiveGraph) -> String {
    let Some(path) = graph.source_path.as_ref() else {
        return "save_graph requires --graph mode (no source path bound).".to_string();
    };
    let path_str = path.to_string_lossy().into_owned();
    // save_disk needs &mut DirGraph; we hold an Arc<DirGraph>. Clone the
    // Arc and unwrap to get exclusive access (cheap when the Arc is
    // already unique, otherwise we do a deep clone of the storage —
    // acceptable for save since it's an explicit operator action).
    let mut dir_arc = graph.kg.dir().clone();
    let dir = std::sync::Arc::make_mut(&mut dir_arc);
    match dir.save_disk(&path_str) {
        Ok(()) => {
            let overview = compute_schema(dir);
            format!(
                "Saved {path_str} ({} nodes, {} edges).",
                overview.node_count, overview.edge_count
            )
        }
        Err(e) => format!("save_graph error: {e}"),
    }
}

File: kglite/mcp_server/tools.py
"""Built-in MCP tools: cypher_query, graph_overview, save_graph.

Mirrors `crates/kglite-mcp-server/src/tools.rs` with the same handler
shapes, status messages, and 15-row inline cap. Differences:

- We call `kglite.KnowledgeGraph.cypher(...)`/`.describe(...)`/`.save(...)`
  directly (kglite Python API). The GIL is released inside `cypher()`
  for execution, so the wrapping overhead is just Python attr-lookup
  and argument marshalling — sub-microsecond per call.
- Mutating queries route to `graph.cypher(...)` too (the Python API
  handles read/write internally, unlike the Rust shim which had to
  branch on `is_mutation_query`).
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import threading
from typing import Any

import kglite
from kglite.mcp_server.csv_http import CsvHttpConfig, write_csv

NO_GRAPH = "No active graph. Pass --graph X.kgl, or activate one via repo_management('org/repo')."


@dataclass
class ActiveGraph:
    """The kglite graph + its source path (for save_graph)."""

    graph: kglite.KnowledgeGraph
    source_path: Path | None


class GraphState:
    """Shared active-graph slot, swappable at runtime (workspace mode
    re-activates the slot when the operator switches repos). The lock
    serialises swap + read access; tool calls hold the read view for
    the duration of the call."""

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._active: ActiveGraph | None = None

    def load_kgl(self, path: Path) -> None:
        graph = kglite.load(str(path))
        with self._lock:
            self._active = ActiveGraph(graph=graph, source_path=path)

    def build_code_tree(self, dir_path: Path) -> None:
        graph = kglite.code_tree.build(str(dir_path))
        with self._lock:
            self._active = ActiveGraph(graph=graph, source_path=None)

    def bind_embedder(self, embedder: Any) -> None:
        """Bind an embedder instance to the active graph. No-op if no
        active graph yet (workspace mode loads the graph later)."""
        with self._lock:
            if self._active is None:
                return
            self._active.graph.set_embedder(embedder)

    def active(self) -> ActiveGraph | None:
        with self._lock:
            return self._active

    def schema(self) -> tuple[int, int] | None:
        a = self.active()
        if a is None:
            return None
        s = a.graph.schema()
        return (s["node_count"], s["edge_count"])


def run_cypher(
    state: GraphState,
    query: str,
    params: dict[str, Any] | None = None,
    csv_http: CsvHttpConfig | None = None,
) -> str:
    """Run a Cypher query against the active graph. Returns the
    rendered tool body — inline 15-row preview by default, CSV (or URL
    when csv_http is configured) when `FORMAT CSV` appears in the
    query."""
    active = state.active()
    if active is None:
        return NO_GRAPH

    try:
        result = active.graph.cypher(query, params=params or {})
    except Exception as e:  # noqa: BLE001 — surface the engine's error verbatim
        return f"Cypher error: {e}"

    # `FORMAT CSV` is recognised inside kglite's cypher() — when set,
    # the call returns a CSV string instead of a ResultView. Use that
    # signal to route URL-vs-inline.
    if isinstance(result, str):
        if csv_http is not None:
            try:
                name = write_csv(csv_http, result)
            except Exception as e:  # noqa: BLE001
                # Best-effort: fall back to inline CSV on write failure.
                return result + f"\n# (csv_http write failed: {e}; CSV inlined)"
            url = csv_http.url_for(name)
            row_count = _count_csv_rows(result)
            return f"FORMAT CSV: {row_count} row(s) written to {url}\nFetch with: curl {url}"
        return result

    return _format_inline(result)


def _count_csv_rows(csv: str) -> int:
    # Strip trailing newline so we don't double-count an empty final line.
    lines = csv.rstrip("\n").splitlines()
    return max(0, len(lines) - 1)  # minus header


def _format_inline(result: Any) -> str:
    """Render a kglite ResultView as a 15-row inline preview. Matches
    the Rust shim's format."""
    rows = list(result)
    total = len(rows)
    if total == 0:
        return "No results."
    columns = result.columns
    header = f"{total} row(s) (showing first 15):\n" if total > 15 else f"{total} row(s):\n"
    out = [header, "\t".join(columns), "\n"]
    for row in rows[:15]:
        # ResultView rows are tuples of values; render each as repr-ish
        # text, tab-separated. Matches the Rust shim's push_value_repr.
        out.append("\t".join(_repr(v) for v in row))
        out.append("\n")
    return "".join(out)


def _repr(v: Any) -> str:
    if v is None:
        return "null"
    if isinstance(v, bool):
        return "true" if v else "false"
    if isinstance(v, (int, float)):
        return str(v)
    if isinstance(v, str):
        return repr(v)
    return str(v)


def run_overview(
    state: GraphState,
    types: list[str] | None,
    connections: Any,
    cypher: Any,
    temp_cleanup_dir: Path | None,
) -> str:
    active = state.active()
    if active is None:
        return NO_GRAPH

    # Wipe the temp dir on bare overview when temp_cleanup is on. "Bare"
    # = no drill-down args. Matches the Rust shim's P4 behaviour.
    if temp_cleanup_dir is not None and types is None and connections is None and cypher is None:
        _wipe_dir(temp_cleanup_dir)

    kwargs: dict[str, Any] = {}
    if types is not None:
        kwargs["types"] = types
    if connections is not None:
        kwargs["connections"] = connections
    if cypher is not None:
        kwargs["cypher"] = cypher
    try:
        return active.graph.describe(**kwargs)
    except Exception as e:  # noqa: BLE001
        return f"graph_overview error: {e}"


def run_save(state: GraphState) -> str:
    active = state.active()
    if active is None:
        return NO_GRAPH
    if active.source_path is None:
        return "save_graph requires --graph mode (no source path bound)."
    try:
        active.graph.save(str(active.source_path))
    except Exception as e:  # noqa: BLE001
        return f"save_graph error: {e}"
    s = state.schema()
    if s is None:
        return f"Saved {active.source_path}."
    nodes, edges = s
    return f"Saved {active.source_path} ({nodes} nodes, {edges} edges)."


def _wipe_dir(dir_path: Path) -> None:
    """Best-effort directory wipe — log on errors, never raise. Matches
    the Rust shim's `wipe_temp_dir` semantics."""
    import logging
    import shutil

    log = logging.getLogger("kglite.mcp_server.tools")
    if not dir_path.is_dir():
        log.debug("temp_cleanup: directory does not exist: %s", dir_path)
        return
    wiped = 0
    for entry in dir_path.iterdir():
        try:
            if entry.is_dir():
                shutil.rmtree(entry)
            else:
                entry.unlink()
            wiped += 1
        except OSError as e:
            log.debug("temp_cleanup: remove %s failed: %s", entry, e)
    if wiped > 0:
        log.info("temp_cleanup: wiped %d entries from %s", wiped, dir_path)
